File: service/app/src/test/java/io/miragon/blueprint/application/service/CancelContractServiceTest.java
package io.miragon.blueprint.application.service;

import static io.miragon.blueprint.domain.leasing.TestObjectBuilder.testLeasingApplication;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import io.miragon.blueprint.application.port.outbound.ContractPort;
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository;
import io.miragon.blueprint.domain.leasing.ApplicationId;
import io.miragon.blueprint.domain.leasing.ContractId;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import java.util.Optional;
import org.junit.jupiter.api.Test;

class CancelContractServiceTest {

    private final LeasingApplicationRepository repository = mock(LeasingApplicationRepository.class);
    private final ContractPort contract = mock(ContractPort.class);
    private final CancelContractService underTest = new CancelContractService(repository, contract);

    @Test
    void cancelContractRevokesTheContractRecordedOnTheApplication() {

        // given: an application that already carries an issued contract
        LeasingApplication application = testLeasingApplication().contractId(new ContractId("CONTRACT-1")).build();
        when(repository.findById(application.id())).thenReturn(Optional.of(application));

        // when: the contract is cancelled as part of the compensation
        underTest.cancelContract(application.id());

        // then: the recorded contract id is revoked in the contract system
        verify(repository).findById(application.id());
        verify(contract).revokeContract(new ContractId("CONTRACT-1"));
        verifyNoMoreInteractions(repository, contract);
    }

    @Test
    void cancelContractFailsWhenTheApplicationIsUnknown() {

        // given: an id the repository cannot resolve
        ApplicationId id = ApplicationId.of("123e4567-e89b-12d3-a456-426614174000");
        when(repository.findById(id)).thenReturn(Optional.empty());

        // when/then: cancellation fails and no contract is touched
        assertThatThrownBy(() -> underTest.cancelContract(id))
                .isInstanceOf(IllegalStateException.class);
        verify(repository).findById(id);
        verifyNoMoreInteractions(repository, contract);
    }

    @Test
    void cancelContractFailsWhenNoContractWasIssuedForTheApplication() {

        // given: an application that never received a contract
        LeasingApplication application = testLeasingApplication().contractId(null).build();
        when(repository.findById(application.id())).thenReturn(Optional.of(application));

        // when/then: cancellation fails and no contract is touched
        assertThatThrownBy(() -> underTest.cancelContract(application.id()))
                .isInstanceOf(IllegalStateException.class);
        verify(repository).findById(application.id());
        verifyNoMoreInteractions(repository, contract);
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/application/service/CancelContractServiceTest.kt
package io.miragon.blueprint.application.service

import io.miragon.blueprint.application.port.outbound.ContractPort
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository
import io.miragon.blueprint.domain.leasing.ApplicationId
import io.miragon.blueprint.domain.leasing.ContractId
import io.miragon.blueprint.domain.leasing.testLeasingApplication
import io.mockk.Runs
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test

class CancelContractServiceTest {

    private val repository = mockk<LeasingApplicationRepository>()
    private val contract = mockk<ContractPort>()
    private val underTest = CancelContractService(repository = repository, contract = contract)

    @Test
    fun `cancelContract revokes the contract recorded on the application`() {

        // given: an application that already carries an issued contract
        val application = testLeasingApplication(contractId = ContractId("CONTRACT-1"))
        every { repository.findById(application.id) } returns application
        every { contract.revokeContract(any()) } just Runs

        // when: the contract is cancelled as part of the compensation
        underTest.cancelContract(application.id)

        // then: the recorded contract id is revoked in the contract system
        verify { repository.findById(application.id) }
        verify { contract.revokeContract(ContractId("CONTRACT-1")) }
        confirmVerified(repository, contract)
    }

    @Test
    fun `cancelContract fails when the application is unknown`() {

        // given: an id the repository cannot resolve
        val id = ApplicationId.of("123e4567-e89b-12d3-a456-426614174000")
        every { repository.findById(id) } returns null

        // when/then: cancellation fails and no contract is touched
        assertThatThrownBy { underTest.cancelContract(id) }
            .isInstanceOf(IllegalStateException::class.java)
        verify { repository.findById(id) }
        confirmVerified(repository, contract)
    }

    @Test
    fun `cancelContract fails when no contract was issued for the application`() {

        // given: an application that never received a contract
        val application = testLeasingApplication(contractId = null)
        every { repository.findById(application.id) } returns application

        // when/then: cancellation fails and no contract is touched
        assertThatThrownBy { underTest.cancelContract(application.id) }
            .isInstanceOf(IllegalStateException::class.java)
        verify { repository.findById(application.id) }
        confirmVerified(repository, contract)
    }
}
